### PR TITLE
[EN-3414] Rework typemarkers

### PR DESCRIPTION
**Pull request by ilya-shablovsky**
Replace string typemarkers with numbers to improve speed of comparison

### CODE REVIEW

File: Bridge/Resources/class.js [hunk 1]
```diff
@@ -1134,6 +1143,32 @@
...
+    // Bridge.define = Bridge.Class.define;
+    // Bridge.definei = Bridge.Class.definei;
+    // Bridge.init = Bridge.Class.init;
+
+        window.__defineTime = 0;
+        window.__defineiTime = 0;
+        window.__initTime = 0;
+
+        Bridge.define = function(...args){
+            const st = performance.now();
```
Comment (vladimir-tikhonov) [inline Bridge/Resources/class.js:1155]: Please use `lunaStartup`, like here - https://github.com/lunalabsltd/integration/blob/develop/pipeline/templates/html/resources/pug/compress.pug#L164. There's no point in measuring performance in network builds

File: Bridge/Resources/interfaces.js [hunk 2]
```diff
@@ -82,13 +82,13 @@
     });
 
     Bridge.define("Bridge.IPromise", {
-        $kind: "interface"
+        $kind: Bridge.Typemarkers.Interface
```
Comment (vladimir-tikhonov) [inline Bridge/Resources/interfaces.js:85]: I'm pretty sure that this change breaks troll.js - https://github.com/lunalabsltd/troll/blob/master/src/js-parser/index.ts#L135

File: Bridge/Resources/array.js [hunk 3]
```diff
@@ -130,7 +130,7 @@
             if (isFn) {
                 var v = defvalue();
 
-                if (!v || (!v.$kind && typeof v !== "object")) {
+                if (!v || (!v.$kind  && typeof v !== "object")) {
```
Comment (vladimir-tikhonov) [inline Bridge/Resources/array.js:133]: nope

File: Bridge/Resources/class.js [hunk 4]
```diff
@@ -310,36 +312,43 @@
...
 
             var isNested = false;
 
-            if (prop.$kind.substring(0, 7) === "nested ") {
+            if (prop.$kind >= Bridge.Typemarkers.NestedOffset) {
                 isNested = true;
-                prop.$kind = prop.$kind.substr(7);
+                prop.$kind = prop.$kind - Bridge.Typemarkers.NestedOffset;
             }
+            // if (Math.floor(prop.$kind / Bridge.Typemarkers.NestedOffset) > 0) {
```
Comment (vladimir-tikhonov) [inline Bridge/Resources/class.js:323]: nope

File: Bridge/Resources/class.js [hunk 5]
```diff
@@ -310,36 +312,43 @@
...
+
 
-            if (prop.$kind === "enum" && !prop.inherits) {
+                if (prop.$kind === Bridge.Typemarkers.Enum && !prop.inherits) {
                 prop.inherits = [System.IComparable, System.IFormattable];
             }
 
-            var rNames = ["fields", "events", "props", "ctors", "methods"],
-                defaultScheme = Bridge.isFunction(prop.main) ? 0 : 1,
+            // var rNames = ["fields", "events", "props", "ctors", "methods"],
```
Comment (vladimir-tikhonov) [inline Bridge/Resources/class.js:333]: nope

File: Bridge/Resources/class.js [hunk 6]
```diff
@@ -1134,6 +1143,32 @@
     Bridge.Class = base;
     Bridge.Class.$queue = [];
     Bridge.Class.$queueEntry = [];
-    Bridge.define = Bridge.Class.define;
-    Bridge.definei = Bridge.Class.definei;
-    Bridge.init = Bridge.Class.init;
+    // Bridge.define = Bridge.Class.define;
```
Comment (vladimir-tikhonov) [inline Bridge/Resources/class.js:1146]: nope

**Review by vladimir-tikhonov — CHANGES_REQUESTED** [review]

File: Bridge/Resources/class.js (hunk 4 above)
Comment (Tokatak) [inline Bridge/Resources/class.js:323]: This looks like a meaningless commit

File: Bridge/Resources/class.js (hunk 5 above)
Comment (Tokatak) [inline Bridge/Resources/class.js:333]: This as well

File: Bridge/Resources/class.js (hunk 6 above)
Comment (Tokatak) [inline Bridge/Resources/class.js:1146]: Forgotten lines

File: Compiler/Translator/Emitter/Blocks/AnonymousTypeCreateBlock.cs [hunk 7]
```diff
@@ -157,7 +157,7 @@ private void GenereateCtor(AnonymousType type)
             this.EnsureComma();
             this.Write(JS.Fields.KIND);
             this.WriteColon();
-            this.WriteScript(TypeKind.Anonymous.ToString().ToLowerInvariant());
+            this.WriteScript((int)TypeKind.Anonymous + 1);
```
Comment (Tokatak) [inline Compiler/Translator/Emitter/Blocks/AnonymousTypeCreateBlock.cs:160]: Why do we use +1 here ?

File: Compiler/Translator/Emitter/Blocks/ClassBlock.cs [hunk 8]
```diff
@@ -364,7 +364,12 @@ protected virtual void WriteKind()
             this.EnsureComma();
             this.Write(JS.Fields.KIND);
             this.WriteColon();
-            this.WriteScript( (isNested ? "nested " : "") + this.TypeInfo.Type.Kind.ToString().ToLowerInvariant());
+            var kindMarker = (int)this.TypeInfo.Type.Kind + 1;
+            if (isNested)
+            {
+                kindMarker += 1000;
```
Comment (Tokatak) [inline Compiler/Translator/Emitter/Blocks/ClassBlock.cs:370]: += Bridge.Typemarkers.NestedOffset
?

**Review by Tokatak — CHANGES_REQUESTED** [review]

File: Compiler/Translator/Emitter/Blocks/AnonymousTypeCreateBlock.cs (hunk 7 above)
Comment (ilya-shablovsky) [reply Compiler/Translator/Emitter/Blocks/AnonymousTypeCreateBlock.cs]: We need to have offset, to make typechecking easier in JS
Anyway, this shoud be moved to constants 👍

File: Compiler/Translator/Emitter/Blocks/ClassBlock.cs (hunk 8 above)
Comment (ilya-shablovsky) [reply Compiler/Translator/Emitter/Blocks/ClassBlock.cs]: Done

File: Bridge/Resources/interfaces.js (hunk 2 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/interfaces.js:85]: Good point, you are absolutely right
Gonna update it too

Comment (ilya-shablovsky) [reply Bridge/Resources/interfaces.js:85]: Here is the PR for troll: https://github.com/lunalabsltd/troll/pull/26

File: Bridge/Resources/class.js (hunk 1 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/class.js]: Removed

File: Bridge/Resources/array.js (hunk 3 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/array.js]: Removed

File: Bridge/Resources/class.js (hunk 4 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/class.js]: Removed

File: Bridge/Resources/class.js (hunk 5 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/class.js]: Removed

File: Bridge/Resources/class.js (hunk 6 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/class.js]: Removed

File: Bridge/Resources/class.js (hunk 4 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/class.js]: Removed

File: Bridge/Resources/class.js (hunk 5 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/class.js]: Removed

File: Bridge/Resources/class.js (hunk 6 above)
Comment (ilya-shablovsky) [reply Bridge/Resources/class.js]: Removed

**Review by vladimir-tikhonov — APPROVED** [review]

**Comment by Tokatak** [comment]
image

**Review by Tokatak — APPROVED** [review]

**Pull request merged**